api/utils/aws/endpoint.go

```diff
@@ -0,0 +1,164 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package aws
+
+import (
+	"net"
+	"strings"
+
+	"github.com/gravitational/trace"
+)
+
+// IsAWSEndpoint returns true if the input URI is an AWS endpoint.
+func IsAWSEndpoint(uri string) bool {
+	// Note that AWSCNEndpointSuffix contains AWSEndpointSuffix so there is no
+	// need to search for AWSCNEndpointSuffix explicitly.
+	return strings.Contains(uri, AWSEndpointSuffix)
+}
+
+// IsRDSEndpoint returns true if the input URI is an RDS endpoint.
+//
+// https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/Aurora.Overview.Endpoints.html
+func IsRDSEndpoint(uri string) bool {
+	return strings.Contains(uri, RDSEndpointSubdomain) &&
+		IsAWSEndpoint(uri)
+}
+
+// IsRedshiftEndpoint returns true if the input URI is an Redshift endpoint.
+//
+// https://docs.aws.amazon.com/redshift/latest/mgmt/connecting-from-psql.html
+func IsRedshiftEndpoint(uri string) bool {
+	return strings.Contains(uri, RedshiftEndpointSubdomain) &&
+		IsAWSEndpoint(uri)
+}
+
+// trimAWSParentDomain removes common AWS endpoint suffixes from the endpoint.
+func trimAWSParentDomain(endpoint string) (string, error) {
+	if strings.HasSuffix(endpoint, AWSEndpointSuffix) {
+		return strings.TrimSuffix(endpoint, AWSEndpointSuffix), nil
+	}
+
+	if strings.HasSuffix(endpoint, AWSCNEndpointSuffix) {
+		return strings.TrimSuffix(endpoint, AWSCNEndpointSuffix), nil
+	}
+
+	return "", trace.BadParameter("endpoint %v is not an AWS endpoint", endpoint)
+}
+
+// ParseRDSURI extracts the identifier and region from the provided RDS URI.
+func ParseRDSURI(uri string) (id, region string, err error) {
+	endpoint, _, err := net.SplitHostPort(uri)
+	if err != nil {
+		return "", "", trace.Wrap(err)
+	}
+	return ParseRDSEndpoint(endpoint)
+}
+
+// ParseRDSEndpoint extracts the identifier and region from the provided RDS
+// endpoint.
+//
+// RDS/Aurora endpoints look like this:
+// aurora-instance-1.abcdefghijklmnop.us-west-1.rds.amazonaws.com
+// aurora-instance-2.abcdefghijklmnop.rds.cn-north-1.amazonaws.com.cn
+func ParseRDSEndpoint(endpoint string) (id, region string, err error) {
```

Comment:
This function is called ParseRDS**Endpoint** but the Redshift one is called ParseRedshift**URI**. I would stick to consistent terminology.

Reply:
For RDS there are two functions `ParseRDSURI` vs `ParseRDSEndpoint`.

URI is passed from api/types with the port. I feel the port is mandatory in that case. 
But there is another place we need RDS endpoint parsing but only have host name without the port. 

Maybe I can do the same for redshift to have two functions?  or maybe I can keep the `ParseRDSEndpoint` version, and do the `SplitHostPort` outside?

What do you recommend?

Reply:
Decided to have a single function `ParseRDSEndpoint` and only do the `SplitHostPort` when `:` is found.